File: static/js/orchestration_dashboard.js
/**
 * Orchestration Dashboard JavaScript
 * Handles the interactive functionality of the steampunk UI
 */

class OrchestrationDashboard {
    constructor() {
        this.dashboard = null;
        this.chatInterface = null;
        this.currentPhase = 'orchestration';
        this.activeAgents = [];
        this.progressBars = {
            research_progress: 0,
            writing_progress: 0,
            validation_progress: 0,
            overall_progress: 0
        };
        this.notifications = [];
        this.chatMessages = [];
        this.documentPreviews = {};
        this.agentStatus = {};

        this.initializeLLMConfig();
        this.loadSavedLLMConfig();

        this.initializeDashboard();
        this.setupEventListeners();
        this.setupLLMEventListeners();
        this.startPeriodicUpdates();
        this.initializeDashboard();
        this.setupEventListeners();
        this.startPeriodicUpdates();
    }

    async initializeDashboard() {
        try {
            // Import the dashboard and chat interface
            const { dashboard, chat_interface } = await import('/static/js/dashboard_integration.js');
            this.dashboard = dashboard;
            this.chatInterface = chat_interface;

            console.log('Orchestration Dashboard initialized successfully');
            this.addNotification('AI Maestro online and ready for collaboration', 'success');
            this.addChatMessage('AI Maestro', 'Greetings, esteemed legal practitioner! I am your AI Maestro, ready to orchestrate the creation of court-ready legal documents. How may I assist you today?', 'ai');

        } catch (error) {
            console.error('Failed to initialize dashboard:', error);
            this.addNotification('Dashboard initialization failed', 'error');
        }
    }

    setupEventListeners() {
        // Chat input handling
        const chatInput = document.getElementById('chat-input');
        const sendButton = document.getElementById('send-message');

        if (chatInput && sendButton) {
            chatInput.addEventListener('keypress', (e) => {
                if (e.key === 'Enter') {
                    this.sendChatMessage();
                }
            });

            sendButton.addEventListener('click', () => {
                this.sendChatMessage();
            });
        }

        // Action button handling
        document.addEventListener('click', (e) => {
            if (e.target.classList.contains('action-btn')) {
                const action = e.target.dataset.action;
                this.handleAction(action);
            }
        });

        // Tab switching
        document.addEventListener('click', (e) => {
            if (e.target.classList.contains('tab-button')) {
                const tabName = e.target.dataset.tab;
                this.switchTab(tabName);
            }
        });

        // Agent node interactions
        document.addEventListener('click', (e) => {
            if (e.target.closest('.agent-node')) {
                const agentNode = e.target.closest('.agent-node');
                const agentName = agentNode.dataset.agent;
                this.handleAgentClick(agentName);
            }
        });
    }

    async sendChatMessage() {
        const chatInput = document.getElementById('chat-input');
        if (!chatInput || !chatInput.value.trim()) return;

        const message = chatInput.value.trim();
        chatInput.value = '';

        // Add user message
        this.addChatMessage('User', message, 'user');

        // Process with AI Maestro
        if (this.chatInterface) {
            try {
                const response = await this.chatInterface.send_user_message(message);
                this.addChatMessage('AI Maestro', response, 'ai');
            } catch (error) {
                console.error('Chat processing error:', error);
                this.addChatMessage('AI Maestro', 'I apologize, but I encountered an error processing your message. Please try again.', 'ai');
            }
        } else {
            // Fallback response
            const response = await this.generateFallbackResponse(message);
            this.addChatMessage('AI Maestro', response, 'ai');
        }
    }

    async generateFallbackResponse(message) {
        const lowerMessage = message.toLowerCase();

        if (lowerMessage.includes('status') || lowerMessage.includes('progress')) {
            return await this.getStatusResponse();
        } else if (lowerMessage.includes('review') || lowerMessage.includes('show')) {
            return "I can help you review documents. Please specify which document you'd like to examine.";
        } else if (lowerMessage.includes('help')) {
            return "I'm here to help orchestrate your legal document creation. I can show you progress, review documents, handle objections, and guide you through the process.";
        } else {
            return "I'm processing your request. How else can I assist with your legal orchestration needs?";
        }
    }

    async getStatusResponse() {
        const overallProgress = this.progressBars.overall_progress;
        const activeAgents = this.activeAgents.length;

        return `**Current Status Report:**

⚙️ **Overall Progress:** ${overallProgress.toFixed(1)}% complete

🔧 **Active Agents:** ${activeAgents} agents currently working
${this.activeAgents.length > 0 ? '• ' + this.activeAgents.join(', ') : '• All agents idle'}

📊 **Progress Breakdown:**
• Research: ${this.progressBars.research_progress.toFixed(1)}%
• Writing: ${this.progressBars.writing_progress.toFixed(1)}%
• Validation: ${this.progressBars.validation_progress.toFixed(1)}%

🎭 **System Status:** ${overallProgress > 0 ? 'Processing in progress' : 'Ready for input'}

Would you like me to elaborate on any specific aspect?`;
    }

    addChatMessage(sender, message, type = 'info') {
        const chatMessages = document.getElementById('chat-messages');
        if (!chatMessages) return;

        const messageDiv = document.createElement('div');
        messageDiv.className = `chat-message ${type}`;
        messageDiv.innerHTML = `<strong>${sender}:</strong> ${message}`;

        chatMessages.appendChild(messageDiv);
        chatMessages.scrollTop = chatMessages.scrollHeight;

        // Store in memory
        this.chatMessages.push({
            timestamp: Date.now(),
            sender,
            message,
            type
        });

        // Keep only last 50 messages
        if (this.chatMessages.length > 50) {
            this.chatMessages = this.chatMessages.slice(-50);
        }
    }

    addNotification(message, type = 'info') {
        const container = document.getElementById('notifications-container');
        if (!container) return;

        const notificationDiv = document.createElement('div');
        notificationDiv.className = `notification-item ${type}`;
        notificationDiv.textContent = message;

        container.appendChild(notificationDiv);

        // Store in memory
        this.notifications.push({
            timestamp: Date.now(),
            message,
            type
        });

        // Keep only last 10 notifications
        if (this.notifications.length > 10) {
            this.notifications = this.notifications.slice(-10);
            // Also remove from DOM
            while (container.children.length > 10) {
                container.removeChild(container.firstChild);
            }
        }

        // Auto-remove after 10 seconds
        setTimeout(() => {
            if (notificationDiv.parentNode) {
                notificationDiv.remove();
            }
        }, 10000);
    }

    async updateProgress(component, progress) {
        this.progressBars[component] = Math.max(0, Math.min(1, progress));

        const progressBar = document.getElementById(`${component}`);
        const progressText = document.getElementById(`${component}-text`);

        if (progressBar) {
            progressBar.style.width = `${progress * 100}%`;
        }

        if (progressText) {
            progressText.textContent = `${Math.round(progress * 100)}%`;
        }

        // Update overall progress
        this.updateOverallProgress();

        // Trigger mechanical effects
        await this.triggerMechanicalEffects(component, progress);
    }

    updateOverallProgress() {
        const totalProgress = Object.values(this.progressBars).reduce((sum, p) => sum + p, 0);
        const overallProgress = totalProgress / Object.keys(this.progressBars).length;

        this.progressBars.overall_progress = overallProgress;
        this.updateProgress('overall_progress', overallProgress);
    }

    async triggerMechanicalEffects(component, progress) {
        // Trigger visual effects based on progress
        if (progress > 0.8) {
            this.showMechanicalEffect('steam-effect');
            this.showMechanicalEffect('gear-animations');
        } else if (progress > 0.5) {
            this.showMechanicalEffect('lever-effect');
        } else if (progress > 0.2) {
            this.showMechanicalEffect('light-effect');
        }
    }

    showMechanicalEffect(effectName) {
        const effect = document.getElementById(effectName);
        if (effect) {
            effect.style.display = 'block';
            setTimeout(() => {
                effect.style.display = 'none';
            }, 3000);
        }
    }

    async activateAgent(agentName) {
        if (!this.activeAgents.includes(agentName)) {
            this.activeAgents.push(agentName);
        }

        this.agentStatus[agentName] = 'active';

        // Update UI
        const agentNode = document.querySelector(`[data-agent="${agentName}"]`);
        if (agentNode) {
            agentNode.classList.add('active');
            const statusElement = agentNode.querySelector('.agent-status');
            if (statusElement) {
                statusElement.textContent = 'ACTIVE';
                statusElement.classList.add('active');
            }
        }

        this.addNotification(`Agent ${agentName} activated`, 'success');
    }

    async deactivateAgent(agentName) {
        if (this.activeAgents.includes(agentName)) {
            this.activeAgents = this.activeAgents.filter(name => name !== agentName);
        }

        this.agentStatus[agentName] = 'idle';

        // Update UI
        const agentNode = document.querySelector(`[data-agent="${agentName}"]`);
        if (agentNode) {
            agentNode.classList.remove('active');
            const statusElement = agentNode.querySelector('.agent-status');
            if (statusElement) {
                statusElement.textContent = 'IDLE';
                statusElement.classList.remove('active');
            }
        }
    }

    async handleAction(action) {
        switch (action) {
            case 'review-outline':
                this.addChatMessage('AI Maestro', 'Opening Skeletal Outline for review. This master document contains your Statement of Facts and Causes of Action. Please review and provide any objections or approvals.', 'ai');
                this.switchTab('skeletal-outline');
                break;

            case 'review-claims':
                this.addChatMessage('AI Maestro', 'Displaying Claims Matrix. This shows all identified legal claims with supporting elements. Review for completeness and accuracy.', 'ai');
                this.switchTab('claims-matrix');
                break;

            case 'review-facts':
                this.addChatMessage('AI Maestro', 'Showing Fact Matrix. This contains all relevant facts organized by category. Check for objectivity and completeness.', 'ai');
                this.switchTab('fact-matrix');
                break;

            case 'approve':
                await this.handleApproval();
                break;

            case 'objection':
                await this.handleObjection();
                break;

            default:
                this.addNotification(`Unknown action: ${action}`, 'warning');
        }
    }

    async handleApproval() {
        this.addChatMessage('AI Maestro', 'Excellent! Your approval has been noted and the system will proceed to the next phase of orchestration.', 'ai');
        this.addNotification('User approval received - proceeding with orchestration', 'success');

        // Update progress
        await this.updateProgress('overall_progress', 1.0);
    }

    async handleObjection() {
        const objection = prompt('Please describe your objection:');
        if (objection) {
            this.addChatMessage('User', `I object to: ${objection}`, 'user');
            this.addChatMessage('AI Maestro', 'I understand your objection and have initiated a research loop to address your concerns. The relevant agents will analyze and update the documents accordingly.', 'ai');
            this.addNotification('Research loop initiated for user objection', 'warning');

            // Trigger research loop effects
            await this.triggerResearchLoop();
        }
    }

    async triggerResearchLoop() {
        await this.activateAgent('rules_of_law');
        await this.activateAgent('issuespotter');
        await this.activateAgent('caselaw_researcher');

        await this.updateProgress('research_progress', 0.1);

        // Simulate research process
        setTimeout(() => this.updateProgress('research_progress', 0.5), 2000);
        setTimeout(() => this.updateProgress('research_progress', 0.8), 4000);
        setTimeout(() => this.updateProgress('research_progress', 1.0), 6000);
        setTimeout(() => this.completeResearchLoop(), 8000);
    }

    async completeResearchLoop() {
        await this.deactivateAgent('rules_of_law');
        await this.deactivateAgent('issuespotter');
        await this.deactivateAgent('caselaw_researcher');

        this.addChatMessage('AI Maestro', 'Research loop completed. Documents have been updated based on your objection. Please review the revised materials.', 'ai');
        this.addNotification('Research loop completed - documents updated', 'success');
    }

    switchTab(tabName) {
        // Hide all tabs
        document.querySelectorAll('.tab-pane').forEach(pane => {
            pane.classList.remove('active');
        });

        // Show selected tab
        const selectedTab = document.getElementById(`${tabName}-content`);
        if (selectedTab) {
            selectedTab.classList.add('active');
        }

        // Update tab buttons
        document.querySelectorAll('.tab-button').forEach(button => {
            button.classList.remove('active');
        });

        const selectedButton = document.querySelector(`[data-tab="${tabName}"]`);
        if (selectedButton) {
            selectedButton.classList.add('active');
        }
    }

    handleAgentClick(agentName) {
        const status = this.agentStatus[agentName] || 'idle';
        const statusText = status === 'active' ? 'Currently active and processing' : 'Currently idle - click to activate';

        this.addChatMessage('AI Maestro', `Agent ${agentName}: ${statusText}`, 'ai');
    }

    startPeriodicUpdates() {
        // Simulate periodic updates for demo purposes
        setInterval(() => {
            this.simulatePeriodicUpdates();
        }, 5000);
    }

    simulatePeriodicUpdates() {
        // Simulate random progress updates for demo
        const components = Object.keys(this.progressBars);
        const randomComponent = components[Math.floor(Math.random() * components.length)];
        const currentProgress = this.progressBars[randomComponent];

        if (currentProgress < 1.0 && Math.random() < 0.3) {
            const newProgress = Math.min(1.0, currentProgress + Math.random() * 0.1);
            this.updateProgress(randomComponent, newProgress);
        }
    }
    // LLM Configuration Methods
    initializeLLMConfig() {
        // LLM Configuration
        this.llmConfig = {
            provider: 'openai',
            model: 'gpt-4o',
            temperature: 0.7,
            maxTokens: 4000
        };
        this.llmProviders = {
            openai: {
                name: 'OpenAI',
                models: ['gpt-4o', 'gpt-4-turbo', 'gpt-3.5-turbo'],
                baseUrl: 'https://api.openai.com/v1'
            },
            anthropic: {
                name: 'Anthropic',
                models: ['claude-3-opus', 'claude-3-sonnet', 'claude-3-haiku'],
                baseUrl: 'https://api.anthropic.com'
            },
            google: {
                name: 'Google',
                models: ['gemini-pro', 'gemini-pro-vision'],
                baseUrl: 'https://generativelanguage.googleapis.com'
            },
            azure: {
                name: 'Azure OpenAI',
                models: ['gpt-4', 'gpt-35-turbo'],
                baseUrl: null // Requires custom endpoint
            },
            local: {
                name: 'Local Model',
                models: ['llama-2-7b', 'llama-2-13b', 'codellama'],
                baseUrl: 'http://localhost:11434'
            }
        };
    }

    setupLLMEventListeners() {
        // LLM Provider change handler
        const providerSelect = document.getElementById('llm-provider');
        if (providerSelect) {
            providerSelect.addEventListener('change', (e) => {
                this.handleProviderChange(e.target.value);
            });
        }

        // LLM Model change handler
        const modelSelect = document.getElementById('llm-model');
        if (modelSelect) {
            modelSelect.addEventListener('change', (e) => {
                this.llmConfig.model = e.target.value;
                this.addNotification(`Model changed to ${e.target.value}`, 'info');
            });
        }

        // Temperature slider handler
        const tempSlider = document.getElementById('llm-temperature');
        const tempValue = document.getElementById('temperature-value');
        if (tempSlider && tempValue) {
            tempSlider.addEventListener('input', (e) => {
                const value = parseFloat(e.target.value);
                this.llmConfig.temperature = value;
                tempValue.textContent = value.toFixed(1);
            });
        }

        // Max tokens input handler
        const maxTokensInput = document.getElementById('llm-max-tokens');
        if (maxTokensInput) {
            maxTokensInput.addEventListener('change', (e) => {
                const value = parseInt(e.target.value);
                if (value >= 100 && value <= 32000) {
                    this.llmConfig.maxTokens = value;
                    this.addNotification(`Max tokens set to ${value}`, 'info');
                } else {
                    e.target.value = this.llmConfig.maxTokens;
                    this.addNotification('Max tokens must be between 100 and 32,000', 'warning');
                }
            });
        }

        // Apply config button handler
        const applyBtn = document.getElementById('apply-llm-config');
        if (applyBtn) {
            applyBtn.addEventListener('click', () => {
                this.applyLLMConfig();
            });
        }

        // Reset config button handler
        const resetBtn = document.getElementById('reset-llm-config');
        if (resetBtn) {
            resetBtn.addEventListener('click', () => {
                this.resetLLMConfig();
            });
        }
    }

    handleProviderChange(provider) {
        this.llmConfig.provider = provider;
        const modelSelect = document.getElementById('llm-model');

        if (modelSelect && this.llmProviders[provider]) {
            // Update available models for the selected provider
            const models = this.llmProviders[provider].models;
            modelSelect.innerHTML = models.map(model =>
                `<option value="${model}">${model}</option>`
            ).join('');

            // Set default model for the provider
            this.llmConfig.model = models[0];
            this.addNotification(`Switched to ${this.llmProviders[provider].name} provider`, 'success');
        }
    }

    async applyLLMConfig() {
        try {
            // Validate configuration
            if (!this.llmConfig.provider || !this.llmConfig.model) {
                throw new Error('Provider and model are required');
            }

            // Update LLM status
            this.updateLLMStatus('Applying configuration...', 'busy');

            // Simulate API call to apply configuration
            await this.delay(1000);

            // Save configuration to localStorage
            localStorage.setItem('lawyerfactory-llm-config', JSON.stringify(this.llmConfig));

            // Update UI
            this.updateLLMStatus('Configuration applied successfully', 'success');
            this.addNotification('LLM configuration applied successfully', 'success');

            // Update chat interface with new config if available
            if (this.chatInterface && this.chatInterface.updateConfig) {
                await this.chatInterface.updateConfig(this.llmConfig);
            }

        } catch (error) {
            console.error('Failed to apply LLM configuration:', error);
            this.updateLLMStatus('Configuration failed', 'error');
            this.addNotification(`Failed to apply configuration: ${error.message}`, 'error');
        }
    }

    resetLLMConfig() {
        // Reset to default configuration
        this.llmConfig = {
            provider: 'openai',
            model: 'gpt-4o',
            temperature: 0.7,
            maxTokens: 4000
        };

        // Update UI elements
        const providerSelect = document.getElementById('llm-provider');
        const modelSelect = document.getElementById('llm-model');
        const tempSlider = document.getElementById('llm-temperature');
        const tempValue = document.getElementById('temperature-value');
        const maxTokensInput = document.getElementById('llm-max-tokens');

        if (providerSelect) providerSelect.value = this.llmConfig.provider;
        if (modelSelect) this.handleProviderChange(this.llmConfig.provider);
        if (tempSlider) tempSlider.value = this.llmConfig.temperature;
        if (tempValue) tempValue.textContent = this.llmConfig.temperature.toFixed(1);
        if (maxTokensInput) maxTokensInput.value = this.llmConfig.maxTokens;

        this.updateLLMStatus('Configuration reset to defaults', 'ready');
        this.addNotification('LLM configuration reset to defaults', 'info');
    }

    updateLLMStatus(message, status) {
        const statusElement = document.getElementById('llm-status-text');
        const indicatorElement = document.getElementById('llm-connection-status');

        if (statusElement) statusElement.textContent = message;

        if (indicatorElement) {
            indicatorElement.className = 'status-indicator';
            switch (status) {
                case 'success':
                    indicatorElement.classList.add('status-online');
                    indicatorElement.textContent = '🟢';
                    break;
                case 'busy':
                    indicatorElement.classList.add('status-busy');
                    indicatorElement.textContent = '🟡';
                    break;
                case 'error':
                    indicatorElement.classList.add('status-error');
                    indicatorElement.textContent = '🔴';
                    break;
                default:
                    indicatorElement.textContent = '⚪';
            }
        }
    }

    loadSavedLLMConfig() {
        try {
            const saved = localStorage.getItem('lawyerfactory-llm-config');
            if (saved) {
                const savedConfig = JSON.parse(saved);
                this.llmConfig = { ...this.llmConfig, ...savedConfig };
                this.addNotification('Loaded saved LLM configuration', 'info');
            }
        } catch (error) {
            console.error('Failed to load saved LLM configuration:', error);
        }
    }

    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    getLLMConfigSummary() {
        const provider = this.llmProviders[this.llmConfig.provider];
        return {
            provider: provider ? provider.name : 'Unknown',
            model: this.llmConfig.model,
            temperature: this.llmConfig.temperature,
            maxTokens: this.llmConfig.maxTokens
        };
    }
}

// Initialize the dashboard when the page loads
document.addEventListener('DOMContentLoaded', () => {
    window.orchestrationDashboard = new OrchestrationDashboard();
});

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = OrchestrationDashboard;
}

File: src/lawyerfactory/phases/07_orchestration/orchestration_dashboard.py
"""
# Script Name: orchestration_dashboard.py
# Description: Steampunk UI Dashboard for LawyerFactory Orchestration Phase.  A whimsical, mechanical-themed interface that provides an immersive AI Maestro partnership experience with three-panel layout:  - Left Panel: User controls, chat interface, progress indicators - Center Panel: Agent network visualization with steampunk aesthetics - Right Panel: Document viewers for real-time preview  Features mechanical animations, sound effects, and interactive elements that make the user feel like they're working with a sophisticated AI machine.
# Relationships:
#   - Entity Type: UI Component
#   - Directory Group: Frontend
#   - Group Tags: user-interface
Steampunk UI Dashboard for LawyerFactory Orchestration Phase.

A whimsical, mechanical-themed interface that provides an immersive
AI Maestro partnership experience with three-panel layout:

- Left Panel: User controls, chat interface, progress indicators
- Center Panel: Agent network visualization with steampunk aesthetics
- Right Panel: Document viewers for real-time preview

Features mechanical animations, sound effects, and interactive elements
that make the user feel like they're working with a sophisticated AI machine.
"""

import logging
from typing import Dict, List, Optional, Any, Callable
from dataclasses import dataclass, field
import json
import asyncio

logger = logging.getLogger(__name__)


@dataclass
class UIData:
    """Data structure for UI state management"""
    current_phase: str = "orchestration"
    active_agents: List[str] = field(default_factory=list)
    progress_bars: Dict[str, float] = field(default_factory=dict)
    notifications: List[str] = field(default_factory=list)
    chat_messages: List[Dict[str, Any]] = field(default_factory=list)
    document_previews: Dict[str, str] = field(default_factory=dict)
    agent_status: Dict[str, str] = field(default_factory=dict)
    mechanical_effects: Dict[str, bool] = field(default_factory=dict)


class OrchestrationDashboard:
    """Main dashboard controller for the steampunk UI"""

    def __init__(self):
        self.ui_data = UIData()
        self.event_handlers: Dict[str, List[Callable]] = {}
        self._initialize_dashboard()

    def _initialize_dashboard(self):
        """Initialize the dashboard with default state"""
        self.ui_data.progress_bars = {
            "research_progress": 0.0,
            "writing_progress": 0.0,
            "validation_progress": 0.0,
            "overall_progress": 0.0
        }

        self.ui_data.agent_status = {
            "rules_of_law": "idle",
            "issuespotter": "idle",
            "citation_formatter": "idle",
            "civil_procedure": "idle",
            "fact_objectivity": "idle",
            "claim_validator": "idle",
            "caselaw_researcher": "idle"
        }

        self.ui_data.mechanical_effects = {
            "gears_spinning": False,
            "steam_hissing": False,
            "levers_pulling": False,
            "lights_flashing": False
        }

        logger.info("Orchestration Dashboard initialized with steampunk theme")

    async def update_progress(self, component: str, progress: float):
        """Update progress bar for a specific component"""
        if component in self.ui_data.progress_bars:
            self.ui_data.progress_bars[component] = max(0.0, min(1.0, progress))
            await self._trigger_mechanical_effects(component, progress)
            await self._emit_event("progress_updated", {"component": component, "progress": progress})

    async def activate_agent(self, agent_name: str):
        """Activate an agent and update UI accordingly"""
        if agent_name not in self.ui_data.active_agents:
            self.ui_data.active_agents.append(agent_name)

        self.ui_data.agent_status[agent_name] = "active"
        await self._trigger_agent_animation(agent_name, "activate")
        await self._emit_event("agent_activated", {"agent": agent_name})

    async def deactivate_agent(self, agent_name: str):
        """Deactivate an agent"""
        if agent_name in self.ui_data.active_agents:
            self.ui_data.active_agents.remove(agent_name)

        self.ui_data.agent_status[agent_name] = "idle"
        await self._trigger_agent_animation(agent_name, "deactivate")
        await self._emit_event("agent_deactivated", {"agent": agent_name})

    async def add_chat_message(self, sender: str, message: str, message_type: str = "info"):
        """Add a message to the chat interface"""
        chat_message = {
            "timestamp": asyncio.get_event_loop().time(),
            "sender": sender,
            "message": message,
            "type": message_type
        }

        self.ui_data.chat_messages.append(chat_message)

        # Keep only last 50 messages
        if len(self.ui_data.chat_messages) > 50:
            self.ui_data.chat_messages = self.ui_data.chat_messages[-50:]

        await self._emit_event("chat_message_added", chat_message)

    async def update_document_preview(self, document_type: str, content: str):
        """Update document preview in the right panel"""
        self.ui_data.document_previews[document_type] = content
        await self._emit_event("document_updated", {"type": document_type, "content": content})

    async def add_notification(self, message: str, notification_type: str = "info"):
        """Add a notification to the UI"""
        notification = f"[{notification_type.upper()}] {message}"
        self.ui_data.notifications.append(notification)

        # Keep only last 10 notifications
        if len(self.ui_data.notifications) > 10:
            self.ui_data.notifications = self.ui_data.notifications[-10:]

        await self._trigger_notification_effect(notification_type)
        await self._emit_event("notification_added", {"message": notification, "type": notification_type})

    async def trigger_research_loop(self, reason: str):
        """Trigger visual effects for research loop activation"""
        await self.add_chat_message("AI Maestro", f"Initiating research loop: {reason}", "warning")
        await self._trigger_mechanical_effects("research_loop", 1.0)
        await self.update_progress("research_progress", 0.1)

    async def complete_phase_step(self, step_name: str):
        """Mark a phase step as completed with visual feedback"""
        await self.add_chat_message("AI Maestro", f"Completed: {step_name}", "success")
        await self._trigger_completion_effect(step_name)

    def get_ui_state(self) -> Dict[str, Any]:
        """Get current UI state for rendering"""
        return {
            "current_phase": self.ui_data.current_phase,
            "active_agents": self.ui_data.active_agents,
            "progress_bars": self.ui_data.progress_bars,
            "notifications": self.ui_data.notifications[-5:],  # Last 5 notifications
            "chat_messages": self.ui_data.chat_messages[-10:],  # Last 10 messages
            "document_previews": self.ui_data.document_previews,
            "agent_status": self.ui_data.agent_status,
            "mechanical_effects": self.ui_data.mechanical_effects
        }

    def on_event(self, event_type: str, handler: Callable):
        """Register event handler"""
        if event_type not in self.event_handlers:
            self.event_handlers[event_type] = []
        self.event_handlers[event_type].append(handler)

    async def _emit_event(self, event_type: str, data: Dict[str, Any]):
        """Emit UI event to registered handlers"""
        if event_type in self.event_handlers:
            for handler in self.event_handlers[event_type]:
                try:
                    await handler(data)
                except Exception as e:
                    logger.error(f"Error in event handler for {event_type}: {e}")

    async def _trigger_mechanical_effects(self, component: str, progress: float):
        """Trigger mechanical visual effects based on progress"""
        if progress > 0.8:
            self.ui_data.mechanical_effects["gears_spinning"] = True
            self.ui_data.mechanical_effects["steam_hissing"] = True
        elif progress > 0.5:
            self.ui_data.mechanical_effects["levers_pulling"] = True
        elif progress > 0.2:
            self.ui_data.mechanical_effects["lights_flashing"] = True

        # Reset effects when progress is low
        if progress < 0.1:
            self.ui_data.mechanical_effects = {k: False for k in self.ui_data.mechanical_effects}

    async def _trigger_agent_animation(self, agent_name: str, action: str):
        """Trigger agent-specific animations"""
        animation_data = {
            "agent": agent_name,
            "action": action,
            "timestamp": asyncio.get_event_loop().time()
        }
        await self._emit_event("agent_animation", animation_data)

    async def _trigger_notification_effect(self, notification_type: str):
        """Trigger visual effects for notifications"""
        if notification_type == "error":
            self.ui_data.mechanical_effects["lights_flashing"] = True
        elif notification_type == "warning":
            self.ui_data.mechanical_effects["steam_hissing"] = True
        elif notification_type == "success":
            self.ui_data.mechanical_effects["gears_spinning"] = True

    async def _trigger_completion_effect(self, step_name: str):
        """Trigger completion effects"""
        await self._emit_event("completion_effect", {"step": step_name})

        # Update overall progress
        completed_steps = sum(1 for p in self.ui_data.progress_bars.values() if p >= 1.0)
        total_steps = len(self.ui_data.progress_bars)
        overall_progress = completed_steps / total_steps if total_steps > 0 else 0
        await self.update_progress("overall_progress", overall_progress)


class SteampunkChatInterface:
    """Chat interface for AI Maestro partnership experience"""

    def __init__(self, dashboard: OrchestrationDashboard):
        self.dashboard = dashboard
        self.conversation_history: List[Dict[str, Any]] = []

    async def send_user_message(self, message: str) -> str:
        """Process user message and get AI response"""
        # Add user message to chat
        await self.dashboard.add_chat_message("User", message, "user")

        # Process the message
        response = await self._process_user_message(message)

        # Add AI response to chat
        await self.dashboard.add_chat_message("AI Maestro", response, "ai")

        return response

    async def _process_user_message(self, message: str) -> str:
        """Process user message and generate appropriate response"""
        message_lower = message.lower()

        # Handle common user interactions
        if any(word in message_lower for word in ["help", "what can you do", "commands"]):
            return self._get_help_message()

        elif any(word in message_lower for word in ["status", "progress", "how are we doing"]):
            return await self._get_status_message()

        elif any(word in message_lower for word in ["review", "check", "look at"]):
            return await self._handle_review_request(message)

        elif any(word in message_lower for word in ["objection", "disagree", "change"]):
            return await self._handle_objection(message)

        elif any(word in message_lower for word in ["approve", "good", "continue"]):
            return await self._handle_approval()

        else:
            return await self._get_general_response(message)

    def _get_help_message(self) -> str:
        """Get help message for users"""
        return """
Greetings, esteemed legal practitioner! I am your AI Maestro, ready to orchestrate the creation of court-ready legal documents.

**Available Commands:**
• "Show me the current status" - View progress and active agents
• "Review [document type]" - Review skeletal outline, claims matrix, or fact matrix
• "I object to [specific issue]" - Raise objections for research loop
• "Approve and continue" - Move to next phase
• "Help" - Show this message

**Current Capabilities:**
• Multi-agent legal analysis with 7 specialized agents
• Real-time document generation and preview
• Interactive research loops for objections
• Bluebook citation formatting
• IRAC methodology implementation
• Court filing compliance checking

How may I assist you in this legal orchestration?
        """

    async def _get_status_message(self) -> str:
        """Get current status message"""
        ui_state = self.dashboard.get_ui_state()

        active_agents = len(ui_state["active_agents"])
        overall_progress = ui_state["progress_bars"]["overall_progress"]

        return f"""
**Current Status Report:**

⚙️ **Overall Progress:** {overall_progress:.1%} complete

🔧 **Active Agents:** {active_agents} agents currently working
• {', '.join(ui_state['active_agents'])}

📊 **Progress Breakdown:**
• Research: {ui_state['progress_bars']['research_progress']:.1%}
• Writing: {ui_state['progress_bars']['writing_progress']:.1%}
• Validation: {ui_state['progress_bars']['validation_progress']:.1%}

🎭 **Mechanical Status:** {'All systems operational' if overall_progress > 0 else 'Initializing systems...'}

Would you like me to elaborate on any specific aspect?
        """

    async def _handle_review_request(self, message: str) -> str:
        """Handle document review requests"""
        message_lower = message.lower()

        if "skeletal outline" in message_lower or "outline" in message_lower:
            return "Opening Skeletal Outline for review. This master document contains your Statement of Facts and Causes of Action. Please review and provide any objections or approvals."
        elif "claims matrix" in message_lower or "claims" in message_lower:
            return "Displaying Claims Matrix. This shows all identified legal claims with supporting elements. Review for completeness and accuracy."
        elif "fact matrix" in message_lower or "facts" in message_lower:
            return "Showing Fact Matrix. This contains all relevant facts organized by category. Check for objectivity and completeness."
        else:
            return "What specific document would you like to review? I can show you the Skeletal Outline, Claims Matrix, or Fact Matrix."

    async def _handle_objection(self, message: str) -> str:
        """Handle user objections"""
        await self.dashboard.trigger_research_loop(f"User objection: {message}")

        return """
I understand your objection and have initiated a research loop to address your concerns. The relevant agents will:

1. 🔍 Analyze your specific objection
2. 📚 Research additional legal authorities
3. ⚖️ Review case law and precedents
4. ✏️ Update the relevant documents
5. 📋 Present revised materials for your review

This ensures your documents reflect the most accurate and favorable legal position. I'll notify you when the research is complete and new materials are ready for review.
        """

    async def _handle_approval(self) -> str:
        """Handle user approval"""
        await self.dashboard.complete_phase_step("User Approval Received")

        return """
Excellent! Your approval has been noted and the system will proceed to the next phase of orchestration.

🎉 **Approval Confirmed**
✅ User review completed
✅ Research loops resolved
✅ Ready for final document assembly

The AI Maestro will now coordinate the final assembly of your court-ready documents. This includes:
• Professional formatting
• Bluebook citation compliance
• Court filing preparation
• Final quality assurance

Would you like to specify any particular formatting preferences or additional requirements before we proceed?
        """

    async def _get_general_response(self, message: str) -> str:
        """Get general conversational response"""
        return f"""
I appreciate you sharing that with me. As your AI Maestro partner, I'm here to ensure every aspect of your legal document orchestration meets the highest standards.

💭 **Processing your input:** {message[:100]}...

If you'd like to:
• **Review documents** - Ask to see the skeletal outline, claims matrix, or fact matrix
• **Raise objections** - Tell me what concerns you have
• **Check status** - Ask about current progress
• **Get help** - Ask what I can do for you

I'm ready to assist with any aspect of the legal orchestration process!
        """


# Global dashboard instance
dashboard = OrchestrationDashboard()
chat_interface = SteampunkChatInterface(dashboard)